language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck.hs

```diff
@@ -0,0 +1,122 @@
+-- | Kind/type inference/checking.
+
+{-# LANGUAGE TypeFamilies  #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Language.PlutusIR.TypeCheck
+    (
+    -- * Configuration.
+      DynamicBuiltinNameTypes (..)
+    , TypeCheckConfig (..)
+    , tccDynamicBuiltinNameTypes
+    , PLC.defConfig
+    , dynamicBuiltinNameMeaningsToTypes
+    -- * Kind/type inference/checking.
+    , inferKind
+    , checkKind
+    , typeOfBuiltinName
+    , inferType
+    , checkType
+    , inferTypeOfProgram
+    , checkTypeOfProgram
+    ) where
+
+import           Language.PlutusCore.Constant
+import qualified           Language.PlutusCore.Core as PLC
+import Language.PlutusIR
+import           Language.PlutusIR.Compiler.Error
+import           Language.PlutusIR.Normalize
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Rename
+import qualified           Language.PlutusCore.TypeCheck as PLC
+import           Language.PlutusIR.TypeCheck.Internal
+import           Language.PlutusCore.Universe
+
+import           Control.Monad.Except
+
+{- Note [Goal of PIR typecheker]
+
+Why we do typechecking:
+
+- The deadcode eliminator (at `Optimizer/Deadcode.hs`) may eliminate ill-typed code, which
+would turn, much to a surprise, an ill-typed program to a well-typed one.
+- The let datatypebinds and/or typebinds introduce new types which may escape from the type of the interm.
+Although they may compile fine to PLC and PLC-typecheck correctly, they are invalid in terms of PIR typechecking.
+This would disallow such programs. See for example `./test/recursion/even3Eval` and the discussion at
+<https://groups.google.com/a/iohk.io/forum/#!msg/plutus/6ycMTngVomc/VKeb00DuHwAJ>
+- The error-messages would be more specialized to lets.
+- We can have more flexible or more strict let syntax rules for truly recursive/nonrecursive bindings
```

Comment:
I don't really understand this point?

Reply:
I meant that it would be easier to check that a let is declared as recursive although it does not have to be: e.g. `let (rec) x = 3 in ...`  could signal a "WARNING/ERROR: you should use nonrec"